### PR TITLE
Bug/fix async decrypt

### PR DESCRIPTION
**Pull request by andrejb**

### CODE REVIEW

File: client/src/leap/soledad/client/encdecpool.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+
+class SyncEncrypterPool(SyncEncryptDecryptPool):
+    """
+    Pool of workers that spawn subprocesses to execute the symmetric encryption
+    of documents to be synced.
+    """
+    # TODO implement throttling to reduce cpu usage??
+    WORKERS = multiprocessing.cpu_count()
+    TABLE_NAME = "docs_tosync"
+    FIELD_NAMES = "doc_id PRIMARY KEY, rev, content"
```
Comment (indradhanush) [inline client/src/leap/soledad/client/encdecpool.py]: Should that be `doc_id, PRIMARY KEY` ? Note the comma.

Comment (kalikaneko) [inline client/src/leap/soledad/client/encdecpool.py]: it looks like, but it's correct:  the field is doc_id and PRIMARY KEY is qualifying it.
nice to have you reading prs again indradhanush :)

**Comment by kalikaneko** [comment]
in `target.py`, the _fetchall method should be moved to use adbapi too: 
``` 
def get_encrypted_doc_from_db(self, doc_id, doc_rev):
...
     1             encr.TABLE_NAME,))
     0         res = self._fetchall(sql, (doc_id, doc_rev))                                                                                                                          
     1         if res:
```

otherwise, you hit the error meskio has reported:

```
IN FETCHALL
syncdb <leap.soledad.client.sqlcipher.SQLCipherConnectionPool instance at 0x7fddc81da950>
2015-05-12 12:31:29 samsara leap.soledad.client.target[17605] ERROR Error while getting encrypted doc from db
2015-05-12 12:31:29 samsara leap.soledad.client.target[17605] ERROR SQLCipherConnectionPool instance has no attribute '__exit__'
Traceback (most recent call last):
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/target.py", line 1171, in sync_exchange
    doc.doc_id, doc.rev)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/target.py", line 1329, in get_encrypted_doc_from_db
    res = self._fetchall(sql, (doc_id, doc_rev))
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/target.py", line 1454, in _fetchall
    with self._sync_db:
AttributeError: SQLCipherConnectionPool instance has no attribute '__exit__'
IN FETCHALL
syncdb <leap.soledad.client.sqlcipher.SQLCipherConnectionPool instance at 0x7fddc81da950>
```

File: client/src/leap/soledad/client/sqlcipher.py [hunk 2]
```diff
@@ -744,33 +705,6 @@ def _get_syncer(self, url, creds=None):
     # Symmetric encryption of syncing docs
     #
 
-    def _encrypt_syncing_docs(self):
```
Comment (kalikaneko) [inline client/src/leap/soledad/client/sqlcipher.py]: this method is removed, but it should be removed from the initialization of SQLCipherU1DBSync (where a call to it in the looping call is still made)

Comment (andrejb) [inline client/src/leap/soledad/client/sqlcipher.py]: Fixed.

**Comment by kalikaneko** [comment]
I'm not getting more exeptions with this branch.

However, there's a weird behavior when syncing. When doing a re-sync (ie, after wiping out all the soledad local databases), keymanager doesn't seem to be able to find the secrets, and therefore it generates them again.

It also gets this error when uploading local material again:

```
2015-05-15 14:22:42 samsara leap.soledad.client.target[22812] DEBUG Soledad sync send status: 5/5
2015-05-15 14:22:42 samsara leap.soledad.client.target[22812] DEBUG Removing encrypted document on sync db: D-3ec7c365814e4da88e54bd91c3e2d5c1
2015-05-15 14:22:43 samsara leap.soledad.client.target[22812] WARNING Soledad sync: error while getting document 1/1: 
2015-05-15 14:22:43 samsara leap.soledad.client.target[22812] WARNING Soledad sync: failing gracefully, will recover on next sync.
2015-05-15 14:22:43 samsara leap.soledad.client.target[22812] ERROR sync failed: cancelling sync threads...
2015-05-15 14:22:43 samsara leap.soledad.client.target[22812] ERROR Soledad sync: cancelled sync threads.
2015-05-15 14:22:43 samsara leap.soledad.client.sync[22812] ERROR Soledad sync error: 
2015-05-15 14:22:43 samsara leap.soledad.client.sync[22812] ERROR Traceback (most recent call last):
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/sync.py", line 185, in _sync
    defer_decryption=defer_decryption)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/target.py", line 1292, in sync_exchange
    defer_decryption=defer_decryption)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/target.py", line 1030, in _get_remote_docs
    raise t.exception
InvalidJSON
```

@andrejb can you try to reproduce this by creating a new account, and then logging out the client, erasing all the local dbs, and logging in again?

File: client/src/leap/soledad/client/http_target.py [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+        def _shutdown_and_finish(res):
+            self._sync_decr_pool.close()
+            return new_generation, new_transaction_id
+
+        d = defer.Deferred()
+        d.addCallback(_shutdown_and_finish)
+
+        def _wait_or_finish():
+            if not self._sync_decr_pool.has_finished():
+                reactor.callLater(
```
Comment (kalikaneko) [inline client/src/leap/soledad/client/http_target.py]: in my tests the sync_decr_pool is never able to finish, I'm debugging what could be the cause.

**Comment by kalikaneko** [comment]
these last commits are moving in a great direction methinks :tada: 
however, I'm still having trouble to get a clean sync. it might be I've got somehow bad data from my previous runs. I'm digging into what can possible be stalling my syncs.

**Comment by kalikaneko** [comment]
in my tests (wipeout soledad db, re-sync and see if keys are found) the synchronization stalls.
It looks that the responsible was ```self._finished = threading.Event()```

This PR fixes this issue.
https://github.com/andrejb/soledad/pull/4

however, there's still a bug in which soledad shutdown leaves the pool unclosed, we shoul think of a way of signalling the interruption.

in general I think the strategy of deferring to a thread the loop for coordinating the async multiprocessing pool defeats the purpose of bypassing the limitations of the GIL.
it is *not* true that sleeping on that thread will not affect the reactor!

however, if it works more or less for now, we can leave further refinements (ie, the removal of all occurrences of the threaded paradigm in this code) for soledad 0.8.

File: client/src/leap/soledad/client/http_target.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+
+        :param response: The body and headers of the response.
+        :type response: tuple(str, dict)
+
+        :return: (new_gen, new_trans_id, number_of_changes, doc_id, rev,
+                 content, gen, trans_id)
+        :rtype: tuple
+        """
+        # decode incoming stream
+        parts = response.splitlines()
```
Comment (kalikaneko) [inline client/src/leap/soledad/client/http_target.py]: for some reason, I'm getting ints as the passed response
```
Traceback (most recent call last):
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/internet/defer.py", line 1105, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/sync.py", line 139, in sync
    defer_decryption=defer_decryption)
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/internet/defer.py", line 1105, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/http_target.py", line 270, in sync_exchange
    defer_decryption=defer_decryption)
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/internet/defer.py", line 1105, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/http_target.py", line 401, in _receive_docs
    number_of_changes, ngen, ntrans = yield d
  File "/home/kali/Virtualenvs/leap-bitmask/local/lib/python2.7/site-packages/Twisted-15.1.0-py2.7-linux-x86_64.egg/twisted/internet/defer.py", line 578, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/http_target.py", line 507, in _insert_received_doc
    self._parse_received_doc_response(response)
  File "/home/kali/leap/soledad/client/src/leap/soledad/client/http_target.py", line 556, in _parse_received_doc_response
    parts = response.splitlines()
AttributeError: 'int' object has no attribute 'splitlines'
```

also, it doesn't make sense the stated type of response, everything points to it being a str, isn't it?

Comment (kalikaneko) [inline client/src/leap/soledad/client/http_target.py]: I had to modify the callback style in order to get it working:

```
diff --git a/client/src/leap/soledad/client/http_target.py b/client/src/leap/soledad/client/http_target.py
index 4e035ca..eec71a0 100644
--- a/client/src/leap/soledad/client/http_target.py
+++ b/client/src/leap/soledad/client/http_target.py
@@ -59,7 +59,12 @@ def _unauth_to_invalid_token_error(failure):
 
 
 def getSoledadPage(*args, **kwargs):
+    def print_content(result):
+        print "GOT PAGE", result
+        return result
+
     d = getPage(*args, **kwargs)
+    d.addCallback(print_content)
     d.addErrback(_unauth_to_invalid_token_error)
     return d
 
@@ -397,7 +402,7 @@ class SoledadHTTPSyncTarget(SyncTarget):
         d = self._receive_one_doc(
             headers, last_known_generation, last_known_trans_id,
             sync_id, 0)
-        d.addCallback(self._insert_received_doc, 1, 1)
+        d.addCallback(lambda doc: self._insert_received_doc(1, 1, doc))
         number_of_changes, ngen, ntrans = yield d
 
         if defer_decryption:
@@ -418,10 +423,9 @@ class SoledadHTTPSyncTarget(SyncTarget):
             d = self._receive_one_doc(
                 headers, last_known_generation,
                 last_known_trans_id, sync_id, received)
-            d.addCallback(
-                self._insert_received_doc,
+            d.addCallback(lambda doc: self._insert_received_doc(
                 received + 1,  # the index of the current received doc
-                number_of_changes)
+                number_of_changes, doc))
             deferreds.append(d)
             received += 1
         results = yield defer.gatherResults(deferreds)
@@ -502,6 +506,8 @@ class SoledadHTTPSyncTarget(SyncTarget):
         :param response: The body and headers of the response.
         :type response: tuple(str, dict)
         """
+        print "INSERT idx/total", idx, total
+        print "RESPONSE>>>", response
         new_generation, new_transaction_id, number_of_changes, doc_id, \
             rev, content, gen, trans_id = \
             self._parse_received_doc_response(response)
```

**Pull request merged**